File: services/api/internal/presentation/mcp/citation.go
package mcp

import (
	"fmt"
	"strings"
	"time"
)

var abntMonths = [...]string{"jan.", "fev.", "mar.", "abr.", "maio", "jun.", "jul.", "ago.", "set.", "out.", "nov.", "dez."}

var saoPaulo = time.FixedZone("-0300", -3*60*60)

type citable struct {
	GazetteID     string
	Title         string
	EditionNumber string
	PublishedAt   time.Time
	IsExtra       bool
	SourceURL     string
	PageStart     int
	PageEnd       int
	Checksum      string
}

func pageFragment(pageStart int) string {
	if pageStart <= 0 {
		return ""
	}
	return fmt.Sprintf("#page=%d", pageStart)
}

func pageRange(start, end int) string {
	switch {
	case start <= 0:
		return ""
	case end <= start:
		return fmt.Sprint(start)
	default:
		return fmt.Sprintf("%d-%d", start, end)
	}
}

func archivedPDFURL(webURL string, c citable) string {
	return fmt.Sprintf("%s/api/v1/gazettes/%s/pdf%s", strings.TrimRight(webURL, "/"), c.GazetteID, pageFragment(c.PageStart))
}

func abntDate(t time.Time) string {
	return fmt.Sprintf("%d %s %d", t.Day(), abntMonths[t.Month()-1], t.Year())
}

func sentence(text string) string {
	return strings.TrimRight(text, " \t\n.:;,") + "."
}

func formatCitation(c citable, webURL string, accessed time.Time) string {
	edition := "ed. " + c.EditionNumber
	if c.EditionNumber == "" {
		edition = "ed. s/n"
	}
	if c.IsExtra {
		edition += " (extra)"
	}
	where := []string{edition, abntDate(c.PublishedAt)}
	if pages := pageRange(c.PageStart, c.PageEnd); pages != "" {
		where = append(where, "p. "+pages)
	}
	return strings.Join([]string{
		fmt.Sprintf("SÃO GONÇALO (RJ). Diário Oficial do Município de São Gonçalo, %s.", strings.Join(where, ", ")),
		sentence(c.Title),
		fmt.Sprintf("Disponível em: <%s%s>.", c.SourceURL, pageFragment(c.PageStart)),
		fmt.Sprintf("Cópia arquivada em: <%s>.", archivedPDFURL(webURL, c)),
		fmt.Sprintf("SHA-256 do PDF: %s.", c.Checksum),
		fmt.Sprintf("Acesso em: %s.", abntDate(accessed.In(saoPaulo))),
	}, " ")
}

File: apps/web/src/citation.ts
import { ActHit, Source } from "./api";
import { SOURCE_NAME } from "./types";

const MONTHS = ["jan.", "fev.", "mar.", "abr.", "maio", "jun.", "jul.", "ago.", "set.", "out.", "nov.", "dez."];

type Citable = Pick<ActHit, "gazette_id" | "title" | "edition_number" | "published_at" | "is_extra"
  | "source_url" | "page_start" | "page_end" | "pdf_sha256" | "source">;

function citationAuthor(source: Source) {
  return source === "diario_camara" ? "SÃO GONÇALO (RJ). Câmara Municipal." : "SÃO GONÇALO (RJ).";
}

export function pageFragment(hit: Pick<Citable, "page_start">) {
  return hit.page_start ? `#page=${hit.page_start}` : "";
}

export function pageLabel(hit: Pick<Citable, "page_start" | "page_end">) {
  if (!hit.page_start) return "";
  if (!hit.page_end || hit.page_end === hit.page_start) return `p. ${hit.page_start}`;
  return `p. ${hit.page_start}-${hit.page_end}`;
}

export function archivedPdfUrl(origin: string, hit: Pick<Citable, "gazette_id" | "page_start">) {
  return `${origin}/api/v1/gazettes/${hit.gazette_id}/pdf${pageFragment(hit)}`;
}

function abntDate(d: Date) {
  return `${d.getDate()} ${MONTHS[d.getMonth()]} ${d.getFullYear()}`;
}

function sentence(text: string) {
  return `${text.replace(/[\s.:;,]+$/, "")}.`;
}

export function formatCitation(hit: Citable, origin: string, accessed: Date) {
  const published = new Date(`${hit.published_at}T12:00:00`);
  const edition = `ed. ${hit.edition_number || "s/n"}${hit.is_extra ? " (extra)" : ""}`;
  const where = [edition, abntDate(published), pageLabel(hit)].filter(Boolean).join(", ");
  return [
    `${citationAuthor(hit.source)} ${SOURCE_NAME[hit.source] ?? SOURCE_NAME.diario_prefeitura}, ${where}.`,
    sentence(hit.title),
    `Disponível em: <${hit.source_url}${pageFragment(hit)}>.`,
    `Cópia arquivada em: <${archivedPdfUrl(origin, hit)}>.`,
    `SHA-256 do PDF: ${hit.pdf_sha256}.`,
    `Acesso em: ${abntDate(accessed)}.`,
  ].join(" ");
}
